src/details/ArborX_AccessTraits.hpp

```diff
@@ -0,0 +1,184 @@
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_ACCESS_TRAITS_HPP
+#define ARBORX_ACCESS_TRAITS_HPP
+
+#include <ArborX_DetailsConcepts.hpp> // is_complete
+#include <ArborX_DetailsTags.hpp>
+#include <ArborX_Point.hpp>
+#include <ArborX_Predicates.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+namespace Traits
+{
+
+struct PrimitivesTag
+{
+};
+
+struct PredicatesTag
+{
+};
+
+// Only a declaration so that existence of a specialization can be detected
+template <typename T, typename Tag, typename Enable = void>
+struct Access;
+
+template <typename View, typename Tag>
+struct Access<View, Tag,
+              typename std::enable_if<Kokkos::is_view<View>::value &&
+                                      View::rank == 1>::type>
+{
+  // Returns a const reference
+  KOKKOS_FUNCTION static typename View::const_value_type &get(View const &v,
+                                                              int i)
+  {
+    return v(i);
+  }
+
+  static typename View::size_type size(View const &v) { return v.extent(0); }
+
+  using memory_space = typename View::memory_space;
+};
+
+template <typename View, typename Tag>
+struct Access<View, Tag,
+              typename std::enable_if<Kokkos::is_view<View>::value &&
+                                      View::rank == 2>::type>
+{
+  // Returns by value
+  KOKKOS_FUNCTION static Point get(View const &v, int i)
+  {
+    return {v(i, 0), v(i, 1), v(i, 2)};
+  }
+
+  static typename View::size_type size(View const &v) { return v.extent(0); }
+
+  using memory_space = typename View::memory_space;
+};
+
+} // namespace Traits
+} // namespace ArborX
+
+namespace ArborX
+{
+namespace Details
+{
+
+// archetypal alias for a 'memory_space' type member in access traits
+template <typename Traits>
+using AccessTraitsMemorySpaceArchetypeAlias = typename Traits::memory_space;
+
+// archetypal expression for 'size()' static member function in access traits
+template <typename Traits>
+using AccessTraitsSizeArchetypeExpression = decltype(
+    Traits::size(std::declval<first_template_parameter_t<Traits> const &>()));
+
+// archetypal expression for 'get()' static member function in access traits
+template <typename Traits>
+using AccessTraitsGetArchetypeExpression = decltype(
+    Traits::get(std::declval<first_template_parameter_t<Traits> const &>(), 0));
+
+// Soon to be removed
+template <typename Traits>
+using decay_result_of_get_t =
+    std::decay_t<detected_t<AccessTraitsGetArchetypeExpression, Traits>>;
+} // namespace Details
+
+namespace Traits
+{
+template <typename Access>
+struct Helper
+{
+  using type = Details::decay_result_of_get_t<Access>;
+  using tag = typename Details::Tag<type>::type;
+};
+} // namespace Traits
+
+namespace Details
+{
+
+template <typename Predicates>
+void check_valid_access_traits(Traits::PredicatesTag, Predicates const &)
```

Comment:
I wrote it as a function because I wanted to take advantage of template argument deduction but this can be revisited later.